services/scheduler/src/main/java/org/mule/service/scheduler/internal/executor/ByCallerThreadGroupPolicy.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.service.scheduler.internal.executor;
+
+import static java.lang.Thread.currentThread;
+import static java.util.Collections.unmodifiableSet;
+
+import org.mule.runtime.core.api.scheduler.exception.SchedulerBusyException;
+
+import java.util.Set;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.ThreadPoolExecutor.AbortPolicy;
+
+/**
+ * Dynamically determines the {@link RejectedExecutionHandler} implementation to use according to the {@link ThreadGroup} of the
+ * current thread.
+ * 
+ * @see AbortPolicy
+ * @see WaitPolicy
+ * 
+ * @since 4.0
+ */
+public final class ByCallerThreadGroupPolicy implements RejectedExecutionHandler {
+
+  private final AbortPolicy abort = new AbortPolicy() {
+
+    @Override
+    public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+      throw new SchedulerBusyException("Task " + r.toString() + " rejected from " + executor.toString());
+    }
+  };
+  private final WaitPolicy wait = new WaitPolicy();
+
+  private final Set<ThreadGroup> waitGroups;
+
+  /**
+   * Builds a new {@link ByCallerThreadGroupPolicy} with the given {@code waitGroups}.
+   * 
+   * @param waitGroups the group of threads for which a {@link WaitPolicy} will be applied. For the rest, an {@link AbortPolicy}
+   *        will be applied.
+   */
+  public ByCallerThreadGroupPolicy(Set<ThreadGroup> waitGroups) {
+    this.waitGroups = unmodifiableSet(waitGroups);
+  }
+
+  @Override
+  public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+    if (currentThread().getThreadGroup() != null && waitGroups.contains(currentThread().getThreadGroup())) {
```

Comment:
Should custom thread pool always wait, how can that we hard-coded without knowing what custom pool is used for? Remember HTTP selectors will be custom soon.  Also are these rules generic enough to apply to everything do you think? Did we say CpuIntensive -> CpuLight should fail, i'm not 100% convinced it should.  If we want a general rule then slower->faster should block and faster->slower should fail probably.